File: app/src/main/java/com/antyzero/smoksmog/ui/dialog/InfoDialog.java
package com.antyzero.smoksmog.ui.dialog;


import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.FragmentManager;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.view.View;

/**
 * For info dialog
 */
public abstract class InfoDialog extends DialogFragment {

    private static final String TAG = InfoDialog.class.getSimpleName();

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        View view = getActivity().getLayoutInflater().inflate(getLayoutId(), null, false);
        initView(view);
        builder.setView(view);
        builder.setPositiveButton(android.R.string.ok, (dialog, which) -> {
            dialog.dismiss();
        });
        return builder.create();
    }

    protected void initView(View view) {
        // override if needed
    }

    @LayoutRes
    protected abstract int getLayoutId();

    public static void show(FragmentManager fragmentManager, Event event) {
        InfoDialog infoDialog;

        try {
            infoDialog = (InfoDialog) event.dialogFragment.newInstance();
        } catch (Exception e) {
            throw new IllegalStateException(
                    "Problem with creating fragment dialog " + event.dialogFragment.getSimpleName(), e);
        }

        infoDialog.show(fragmentManager, TAG);
    }

    public static class Event<T extends InfoDialog> {

        @NonNull
        private final Class<T> dialogFragment;

        public Event(@NonNull Class<T> dialogFragment) {
            this.dialogFragment = dialogFragment;
        }
    }
}

File: app/src/main/kotlin/com/antyzero/smoksmog/ui/dialog/InfoDialog.kt
package com.antyzero.smoksmog.ui.dialog


import android.app.AlertDialog
import android.app.Dialog
import android.app.DialogFragment
import android.app.FragmentManager
import android.os.Bundle
import android.view.View

/**
 * For info dialog
 */
abstract class InfoDialog : DialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle): Dialog {
        val builder = AlertDialog.Builder(activity)
        val view = activity.layoutInflater.inflate(getLayoutId(), null, false)
        initView(view)
        builder.setView(view)
        builder.setPositiveButton(android.R.string.ok) { dialog, which -> dialog.dismiss() }
        return updateBuilder(builder).create()
    }

    /**
     * Override to change builder settings

     * @param builder
     * *
     * @return
     */
    protected open fun updateBuilder(builder: AlertDialog.Builder): AlertDialog.Builder {
        return builder
    }

    protected open fun initView(view: View) {
        // override if needed
    }

    protected abstract fun getLayoutId(): Int

    class Event<T : InfoDialog>(internal val dialogFragment: Class<T>)

    companion object {

        private val TAG = InfoDialog::class.java.simpleName

        fun show(fragmentManager: FragmentManager, event: Event<*>) {
            val infoDialog: InfoDialog

            try {
                infoDialog = event.dialogFragment.newInstance() as InfoDialog
            } catch (e: Exception) {
                throw IllegalStateException(
                        "Problem with creating fragment dialog " + event.dialogFragment.simpleName, e)
            }

            infoDialog.show(fragmentManager, TAG)
        }
    }
}
